thunder/dev_utils/debug_transform.py

```diff
@@ -61,10 +64,35 @@ def _post_call_ctx(bsym, *args, **kwargs):
                 post_debug_bsym = create_debug_boundsymbol(post_debug_name, bsym, _post_call_ctx)
                 new_bsyms.append(post_debug_bsym)
 
+            debug_counter += 1
+
         debug_trace.bound_symbols = new_bsyms
         elapsed_time_ns = time.perf_counter_ns() - start_time_ns
 
         debug_trace.set_provenance(TraceProvenance(f"Debug trace (took {elapsed_time_ns * 1e-6:.2f} milliseconds)"))
 
-        debug_counter += 1
         return debug_trace
+
+
+def debug_execution_trace(cfn, pre_callback: Callable | None = None, post_callback: Callable | None = None):
+    """
+    Adds a debugging transform to the trace allowing pre and post execution callbacks.
+
+    The function inserts debug symbols in the computation traces to call the callbacks before and/or after each symbol
+    in the trace. These callbacks can be used to inspect or log information about the execution of the computation.
+
+    Args:
+        cfn: `thunder.jit` function to debug.
+        pre_callback: An optional callable that is executed before each bound symbol is processed.
+            It should have the signature `(BoundSymbol, *args, **kwargs)` and return a string. If `None`, no
+            pre-execution callback is used.
+        post_callback: An optional callable that is executed after each bound symbol is processed.
+            It should have the signature `(BoundSymbol, *args, **kwargs)` and return a string. If `None`, no
+            post-execution callback is used.
+    """
+    if pre_callback is None and post_callback is None:
+        raise RuntimeError(
+            "debug_execution_trace: Both `pre_callback` and `post_callback` were both None, expected atleast one of them to be None."
+        )
+    _debug_tfms = _DebugTransform(pre_callback=pre_callback, post_callback=post_callback)
```

Comment:
would this "_debug_tfms" need "s" which seems to indicate it's plural?

Reply:
Agreed, will just rename it to `_debug_transform` to avoid confusion.